app/models/manageiq/providers/vmware/network_manager/refresh_parser.rb

```diff
@@ -28,36 +28,70 @@ def ems_inv_to_hashes
-      process_collection(@inv[:networks], :cloud_subnets, false) { |n| parse_network_subnet(n) }
+      $vcloud_log.info("#{log_header} Fetched #{@inv[:vdc_networks].count} VDC networks")
     end
 
-    def get_network_ports
-      # TODO: Update FOG so that it will provide list of network ports. Currently, it is only possible to obtain
-      # very little data, barely enough to tell to which network the VM is connected to.
-      @inv[:vm_networks] = []
-      @org.vdcs.each do |vdc|
-        vdc.vapps.each do |vapp|
-          vapp.vms.each do |vm|
-            @inv[:vm_networks] += vm.network_adapters.each_with_index.map do |net_if, i|
-              net_if[:vapp] = vapp
-              net_if[:vm] = vm
-              net_if[:vm_network] = vm.network
-              net_if[:idx] = i
-              net_if
-            end
+    def get_vapp_networks
+      @inv[:vapp_networks] = []
+      @inv[:routers] = []
+
+      @ems.orchestration_stacks.each do |stack|
+        fetch_network_configurations_for_vapp(stack.ems_ref).map do |net_conf|
+          $vcloud_log.debug("#{log_header} processing net_conf for vapp #{stack.ems_ref}: #{net_conf}")
+          network_id = network_id_from_links(net_conf)
+          $vcloud_log.debug("#{log_header} calculated vApp network id: #{network_id}")
+          if (vdc_net = corresponding_vdc_network(net_conf, @inv[:vdc_networks_idx]))
+            $vcloud_log.debug("#{log_header} skipping VDC network duplicate")
+            memorize_network_name_mapping(stack.ems_ref, vdc_net.name, vdc_net.id)
+            next
+          else
+            memorize_network_name_mapping(stack.ems_ref, net_conf[:networkName], network_id)
+            @inv[:vapp_networks] << build_vapp_network(stack, network_id, net_conf)
+          end
+
+          # routers connecting vApp networks to VDC networks
+          if (parent_net = parent_vdc_network(net_conf, @inv[:vdc_networks_idx]))
+            $vcloud_log.debug("#{log_header} connecting router to parent: #{parent_net}")
+            @inv[:routers] << {
+              :net_conf   => net_conf,
+              :network_id => network_id,
+              :parent_net => parent_net
+            }
           end
         end
       end
 
-      process_collection(@inv[:vm_networks], :network_ports) { |n| parse_network_port(n) }
+      process_collection(@inv[:vapp_networks], :cloud_networks) { |n| parse_network(n) }
+      process_collection(@inv[:vapp_networks], :cloud_subnets, false) { |n| parse_network_subnet(n) }
+      process_collection(@inv[:routers], :network_routers) { |r| parse_network_router(r) }
+
+      $vcloud_log.info("#{log_header} Fetched #{@inv[:vapp_networks].count} vApp networks")
+      $vcloud_log.info("#{log_header} Fetched #{@inv[:routers].count} network routers")
+    end
+
+    def get_network_ports
+      @inv[:nics] = []
+      @ems.vms.each do |vm|
+        fetch_nic_configurations_for_vm(vm.ems_ref).each do |nic|
```

Comment:
Is it possible to get all NICs up front instead of one-off API calls per-VM?  That way it won't slow down dramatically as the number of VMs increases (same comment for `fetch_network_configurations_for_vapp`)

Reply:
If we can avoid an N+1 call we should definetely go for it.

Reply:
VMware network refresh parser is so complicated because VMware API is surprisingly modest when it comes to networking. In other words, I wasn't able to find a better way to obtain networking information other than looping through VMs and querying for their networking configuration. API documentation that I was looking at is here: https://www.vmware.com/support/vcd/doc/rest-api-doc-1.5-html.

I'd feel much better if there were API calls like "list all networks", "list all subnets", "list all NICs", "list all Routers", "list all floating IPs" ... but they just aren't there. I'm not sure how anyone was ever able to use the API to reconstruct networking information for vCD to be honest 😄 

Is it possible that VMware API is so modest with information, or did I miss something big?

Reply:
Okay that is unfortunate, is this data that is usually returned with the VM but since we have to split up the parser by manager (cloud vs network) we have to get it again?
Asking because we're planning on unifying the managers into a single provider and single refresh so this issue might go away.

Reply:
Yes, networking data is generally fetched together with VM data. But on this PR we paid extra attention to only fetch networking data for each VM (without other stuff about VM that we assume is already fetched by cloud provider), that's why parsing is a bit complicated (reading from XML directly).

Are you referring to graph refresh? We plan on implementing it very soon (probably as soon as this week).

Reply:
Not really related to graph refresh, although that sounds great.  What we want is to unify all of the managers so instead of having 2 or 3 different workers and different refreshes this could all be done in one so we wouldn't have to fetch the data again just because this is in a different manager